flang/lib/Lower/Intrinsics.cpp

```diff
@@ -589,8 +589,9 @@ mlir::Value IntrinsicLibrary::genRuntimeCall(llvm::StringRef name,
     for (mlir::Value arg : args) {
       auto actualType = actualFuncType.getInput(i);
       if (soughtFuncType.getInput(i) != actualType) {
-        auto castedArg = builder.createHere<fir::ConvertOp>(actualType, arg);
-        convertedArguments.push_back(castedArg.getResult());
+        auto castedArg =
+            builder.convertOnAssign(builder.getLoc(), actualType, arg);
```

Comment:
There are about a dozen or so files that have strange conversion issues. INTEGER -> COMPLEX, COMPLEX -> REAL, etc. They appear to be masked by the front-end (i.e. unreliable conversions in Ev::Expression). Specifically, the expression is assuming characteristics about internal conversions that simply do not hold with respect to LLVM/MLIR/FIR.

Reply:
But is there such integer to complex conversion happening here ? I could not hit any such case. It's a bug in `Intrinsics.cpp` independently of what the front-end does if there is such need here.

Reply:
It looks like I deleted that log file.

According to 10.2 (see table 10.8, etc.), using any sort of numeric type variable conforms with the other numeric types. 16.9.45p4 (degenerate case), allows conversion of an integer to a real. There is no operation for that in LLVM/MLIR/FIR and it involves performing several primitive steps: int2fp, constant-fp, constructing a pair.

I should rename this from `convertOnAssign` to something more generic, but lowering has to construct the operational semantics of the Fortran numeric conversions as they aren't trivial casts.

Reply:
OK, but here (in `genRuntimeCall`) the code is just trying to plug some math intrinsic call into some runtime call and is not really related to any Fortran rules. If you had to fix something here, then it would be a sign that the compiler is lowering something like `acos(cplx)` by inserting a call to `acos(real)` + "conversions". That is logically bad code that should not be generated.
If this change is required for you then it is likely only transforming a compiling error into a logical error in the compiled program (which in my opinion is worse).
Hence, I would rather keep the original code here, find a reproducer of the issue you had and fix the source of the issue, not the symptoms.

Reply:
Is this legal Fortran?
```Fortran
  INTEGER :: i
  x = CONJG(CMPLX(i))
```

This is _not_ legal FIR:
```mlir
  %i_val = fir.load %i : !fir.ref<i32>
  %22 = fir.convert %i_val : (i32) -> !fir.complex<8>
  %x = fir.call @runtime_conjg(%22) : (!fir.complex<8>) -> !fir.complex<8>
```

It's hard to see where `Ev::Expression` is going to insert some extra semantics between the CMPLX operation and the CONJG call. What would it be?

Reply:
Yes, this is legal Fortran.
Weird, I do not get this lowering a with the current bbc:
```
 INTEGER :: i
  x = CONJG(CMPLX(i))
end
```

bbc -emit-fir:

```
func @_QQmain() {
  %0 = fir.alloca i32 {name = "i"}
  %1 = fir.alloca f32 {name = "x"}
  %2 = fir.load %0 : !fir.ref<i32>
  %3 = fir.convert %2 : (i32) -> f32
  %cst = constant 0.000000e+00 : f32
  %4 = fir.undefined !fir.complex<4>
  %c0_i32 = constant 0 : i32
  %5 = fir.insert_value %4, %3, %c0_i32 : (!fir.complex<4>, f32, i32) -> !fir.complex<4>
  %c1_i32 = constant 1 : i32
  %6 = fir.insert_value %5, %cst, %c1_i32 : (!fir.complex<4>, f32, i32) -> !fir.complex<4>
  %c1_i32_0 = constant 1 : i32
  %7 = fir.extract_value %6, %c1_i32_0 : (!fir.complex<4>, i32) -> f32
  %8 = fir.negf %7 : f32
  %c1_i32_1 = constant 1 : i32
  %9 = fir.insert_value %6, %8, %c1_i32_1 : (!fir.complex<4>, f32, i32) -> !fir.complex<4>
  %c0_i32_2 = constant 0 : i32
  %10 = fir.extract_value %9, %c0_i32_2 : (!fir.complex<4>, i32) -> f32
  %11 = fir.convert %10 : (f32) -> f32
  fir.store %11 to %1 : !fir.ref<f32>
  return
}
```
I am building with clang9 + libc++, my head is 6be2b9a53e7480fa155d8f684adac1feaf18b141.

Are you literally seeing a call to `runtime_conjg` ? I do not see which piece of code can generate that... 
"CMPLX" never hits the bridge, it is rewritten to `ComplexConstrutor` by expression semantics (in fact, there is not even any code handling "cmplx" in the intrinsic lowering utility).
This happens here: 
https://github.com/flang-compiler/f18-llvm-project/blob/1f8790050b0e99e7b46cc69518aa84f46f50738e/flang/lib/Evaluate/fold-complex.cpp#L41
https://github.com/flang-compiler/f18-llvm-project/blob/84a099df05320bf7a3c5d97bcc51d53012f24667/flang/include/flang/Evaluate/tools.h#L363

That is also the only way to rewrite it because the typed expression format do not define Convert<T> as a variant for complex expression:
https://github.com/flang-compiler/f18-llvm-project/blob/84a099df05320bf7a3c5d97bcc51d53012f24667/flang/include/flang/Evaluate/expression.h#L563-L564


"CONJG" do hit the bridge, but is handled inline by genConjg here:
https://github.com/flang-compiler/f18-llvm-project/blob/30f2801b59d3bebb945c7da94cff6ba6f632e3ed/flang/lib/Lower/Intrinsics.cpp#L674
There is even a regression test for it:
https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/test/Lower/intrinsics.f90#L49

And there is no code that should ever generate a call to "runtime_conjg" in the bridge (unless you want to outline intrinsics with the flag -outine-intrinsics, in which case it is named fir.conjg.z4.z4, not runtime_conjg) I had not luck grepping for it, though it may be generated somewhere.


So I am very lost here, it may be something in LAPACK code that matters though to reproduce the issue, which file is the issue coming from, zhemm.f ? I could compile it before the boolean change (#32), now I am getting "'loop.if' op operand #0 must be 1-bit signless integer, but got '!fir.logical<4>'" error, but nothing regarding a fir.convert issue.

Reply:
I sent you down the wrong rabbit hole. That was just an off-the-cuff example until I could recreate the list of issues. Then I got distracted with other issues. My sincere apologies.

I still owe you the list of real world examples.

Reply:
I had to rewind to here. https://github.com/schweitzpgi/f18-llvm-project/commit/d91f76969ac23880700e067606c36e54a6876cdf

When I build that and compile BLAS, I get ConvertOp failures in:
snrm2.f, srotg.f, dnrm2.f, drotg.f, scnrm2.f, crotg.f, dcabs1.f, dznrm2.f, and zrotg.f

The failures are because of the following bono-instructions.
```mlir
            %63 = fir.convert %62 : (f32) -> !fir.complex<8>
            %64 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
            %66 = fir.convert %65 : (!fir.complex<8>) -> f32
            %63 = fir.convert %62 : (f32) -> !fir.complex<8>
            %64 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
            %66 = fir.convert %65 : (!fir.complex<8>) -> f32

    %24 = fir.convert %23 : (f32) -> !fir.complex<8>
    %25 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
    %27 = fir.convert %26 : (!fir.complex<8>) -> f32
    %32 = fir.convert %31 : (f32) -> !fir.complex<8>
    %33 = fir.convert %c2_i32_0 : (i32) -> !fir.complex<8>
    %35 = fir.convert %34 : (!fir.complex<8>) -> f32

           %63 = fir.convert %62 : (f64) -> !fir.complex<8>
            %64 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
            %66 = fir.convert %65 : (!fir.complex<8>) -> f64
            %63 = fir.convert %62 : (f64) -> !fir.complex<8>
            %64 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
            %66 = fir.convert %65 : (!fir.complex<8>) -> f64

   %24 = fir.convert %23 : (f64) -> !fir.complex<8>
    %25 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
    %27 = fir.convert %26 : (!fir.complex<8>) -> f64
    %32 = fir.convert %31 : (f64) -> !fir.complex<8>
    %33 = fir.convert %c2_i32_0 : (i32) -> !fir.complex<8>
    %35 = fir.convert %34 : (!fir.complex<8>) -> f64

          %63 = fir.convert %62 : (f32) -> !fir.complex<8>
            %64 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
            %66 = fir.convert %65 : (!fir.complex<8>) -> f32
            %63 = fir.convert %62 : (f32) -> !fir.complex<8>
            %64 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
            %66 = fir.convert %65 : (!fir.complex<8>) -> f32

   %33 = fir.convert %32 : (f32) -> !fir.complex<8>
    %34 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
    %36 = fir.convert %35 : (!fir.complex<8>) -> f32
    %52 = fir.convert %51 : (f32) -> !fir.complex<8>
    %53 = fir.convert %c2_i32_16 : (i32) -> !fir.complex<8>
    %55 = fir.convert %54 : (!fir.complex<8>) -> f32

  %2 = fir.convert %1 : (!fir.complex<8>) -> f64

       %65 = fir.convert %64 : (!fir.complex<8>) -> f64
          %75 = fir.convert %74 : (f64) -> !fir.complex<8>
          %76 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
          %78 = fir.convert %77 : (!fir.complex<8>) -> f64
          %75 = fir.convert %74 : (f64) -> !fir.complex<8>
          %76 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
          %78 = fir.convert %77 : (!fir.complex<8>) -> f64
         %75 = fir.convert %74 : (f64) -> !fir.complex<8>
          %76 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
          %78 = fir.convert %77 : (!fir.complex<8>) -> f64
          %75 = fir.convert %74 : (f64) -> !fir.complex<8>
          %76 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
          %78 = fir.convert %77 : (!fir.complex<8>) -> f64

    %28 = fir.convert %27 : (f64) -> !fir.complex<8>
    %29 = fir.convert %c2_i32 : (i32) -> !fir.complex<8>
    %31 = fir.convert %30 : (!fir.complex<8>) -> f64
    %42 = fir.convert %41 : (f64) -> !fir.complex<8>
    %43 = fir.convert %c2_i32_14 : (i32) -> !fir.complex<8>
    %45 = fir.convert %44 : (!fir.complex<8>) -> f64
```
This patch fixes all of these issues.

Reply:
Thanks for finding back the BLAS issue. I am ok with keeping this `convertOnAssign` here until I find why it's needed.